router/pkg/config/config.go

```diff
@@ -191,6 +192,36 @@ type RequestHeaderRule struct {
 	Default string `yaml:"default"`
 }
 
+type ResponseHeaderRuleAlgorithm string
+
+const (
+	// ResponseHeaderRuleAlgorithmFirstWrite propagates the first response header from a subgraph to the client
+	ResponseHeaderRuleAlgorithmFirstWrite ResponseHeaderRuleAlgorithm = "first_write"
+	// ResponseHeaderRuleAlgorithmLastWrite propagates the last response header from a subgraph to the client
+	ResponseHeaderRuleAlgorithmLastWrite ResponseHeaderRuleAlgorithm = "last_write"
+	// ResponseHeaderRuleAlgorithmAppend appends all response headers from all subgraphs to a comma separated list of values in the client response
+	ResponseHeaderRuleAlgorithmAppend ResponseHeaderRuleAlgorithm = "append"
+	// ResponseHeaderRuleAlgorithmMostRestrictiveCacheControl propagates the most restrictive cache control header from all subgraph responses to the client
+	ResponseHeaderRuleAlgorithmMostRestrictiveCacheControl ResponseHeaderRuleAlgorithm = "most_restrictive_cache_control"
+)
+
+type ResponseHeaderRule struct {
+	// Operation describes the header operation to perform e.g. "propagate"
+	Operation HeaderRuleOperation `yaml:"op"`
+	// Matching is the regex to match the header name against
+	Matching string `yaml:"matching"`
+	// Named is the exact header name to match
+	Named string `yaml:"named"`
+	// Rename renames the header's key to the provided value
+	Rename string `yaml:"rename,omitempty"`
+	// Default is the default value to set if the header is not present
+	Default string `yaml:"default"`
+	// Algorithm is the algorithm to use when multiple headers are present
+	Algorithm ResponseHeaderRuleAlgorithm `yaml:"algorithm,omitempty"`
+	// OnlyEnableOnClientRequestGet only propagates the header if the client request is a GET request
+	OnlyEnableOnClientRequestGet bool `yaml:"only_enable_on_client_request_get,omitempty"`
```

Comment:
@jensneuse  @StarpTech what should I do here, should they be able to specify the propagation of the rule based on the HTTP method? Only for Gets?

Reply:
From my POV, it only makes sense to propagate Cache Control Headers when the client request method is GET, which our Router already supports. However, it might be the case that someone uses a proxy, like Cloudflare Worker to rewrite POST to GET, so we could also allow them to pass CC headers on non GET. But the default should be to only do it for GET. Alternatively, we could also allow it for all methods and limit a rule to certain methods if specified.

Reply:
Sounds reasonable.